pages/api/zapbye.ts

```diff
@@ -0,0 +1,100 @@
+      const {
+            method,
+      } = req;
+
+      if (method != "POST") {
+            res.statusCode = 405;
+            res.json({ error: "Method Not Allowed" });
+            return;
+      }
+
+      if(!salesforceConnection || !salesforceConnection.accessToken) {
+            console.log("Attempting to log in!");
+            salesforceConnection = new jsforce.Connection();
+            await salesforceConnection.login(CONSTANTS.SALESFORCE_EMAIL, CONSTANTS.SALESFORCE_PASSWORD + CONSTANTS.SALESFORCE_SECURITY_TOKEN, function(err, res) {
+                  if (err) {
+                        salesforceConnection = null;
+                        res.statusCode = 500;
+                        res.body = err;
+                        return console.error(err);
+                  }
+            });
+      }
+      let lemlistPayload = req.body;
+      await salesforceConnection.sobject(CONSTANTS.SALESFORCE_SOBJECT_TYPE).create({
+            TaskSubtype: CONSTANTS.SALESFORCE_TASK_SUBTYPE,
+            Subject: CONSTANTS.SALESFORCE_TASK_SUBJECT,
+            Status: CONSTANTS.SALESFORCE_STATUS,
+            RecordTypeId: CONSTANTS.SALESFORCE_LEMLISTTASK_ID,
+            ActivityDate: new Date(), //DueDate
+            Lemlist_Campaign_Name__c: await getLemlistCampaignNameFromCampaignId(lemlistPayload.campaignId),
+            Lemlist_Sender__c: lemlistPayload.leadFirstName + ' ' + lemlistPayload.leadLastName,
+            Lemlist_Email__c: lemlistPayload.leadEmail,
+            Lemlist_Task_Type__c: lemlistPayload.type,
+            Lemlist_Sequence_Step_Number__c: lemlistPayload.sequenceStep,
+      },{},function(err, ret) {
+            if (err || !ret.success) {
+                  return console.error(err, ret);
+            }
+            console.log('Task created successfully: ', ret);
+      });
+      res.statusCode = 200;
+      res.json({ response: "200 OK" });
+}
+
+interface LemlistCampaign {
+      _id: string,
+      name: string,
+      archived: boolean
+}
+
+async function getLemlistCampaigns(): Promise<JSON> {
+      let endpoint = "https://:" + CONSTANTS.LEMLIST_API_KEY + "@api.lemlist.com/api/campaigns";
+      console.log("Attempting to fetch Lemlist Campaigns from API");
+      const response: Response = await fetch(endpoint);
+      return response.json();
+}
+
+async function getLemlistCampaignNameFromCampaignId(campaignId: string): Promise<string> {
+      let campaign = filterCampaignById(cachedCampaignsArray, campaignId);
+      if(!campaign) {
+            //we keep track of how long ago we last pulled from Lemlist so we don't end up getting rate limited if there's an ID we can't find
+            let now = Date.now();
+            if(!lastCampaignCheckedTimestamp || now - lastCampaignCheckedTimestamp > CONSTANTS.LEMLIST_CAMPAIGN_FETCH_COOLDOWN_TIME_IN_MS) {
+                  cachedCampaignsArray = await getLemlistCampaigns() as unknown as Array<LemlistCampaign>;
```

Comment:
Someone must know the correct way to do this. HALP!

Reply:
I think setting a type in the declaration that matches what the function returns, will do the job.

Reply:
Yeah you're right, instead of returning Promise<JSON> in the method you can define an interface the shape of the data that you want. Since Typescript is a pretty loose typescript, you don't even have to define everything that's being returned, you can literally just pick the elements you want. Here's an [example](https://github.com/opticpower/billing/blob/master/pages/api/pandadoc.ts#L40) where I defined a PandadockWebhook type and just loosely described the shape of the data I wanted so I could use it. The actual webhook returns a lot more data, but I didn't need it for my purpose so I didn't type it.

Reply:
@mattvv I see in your method you get it from the request coming in, but I get mine from an HTTP request via fetch. The only 2 options I see is getting JSON via `.json()` or getting a ReadableStream via `.body()`. I can have it `GetLemlistCampaigns` returns the proper type, but with what I know now I would just be moving the hacky syntax to a different function.

Reply:
I think the .json() method returns a Promise<any> allowing you to cast to anything.

Reply:
Wow ok, I just figured out how to do this. I had to pull it out of the Promise first with the away and then cast it after that. Ended up looking like this `return await response.json() as Array<LemlistCampaign>;`